packages/eslint-plugin/tests/rules/promise-function-async.test.ts

```diff
@@ -788,5 +826,60 @@ async function promiseInUnionWithoutExplicitReturnType(p: boolean) {
 }
       `,
     },
+    {
+      code: `
+function overloadingThatCanReturnPromise(): Promise<number>;
+function overloadingThatCanReturnPromise(a: boolean): Promise<string>;
+function overloadingThatCanReturnPromise(
+  a?: boolean,
+): Promise<number | string> {
+  return Promise.resolve(5);
+}
+      `,
+      errors: [
+        {
+          messageId,
+        },
+      ],
+      output: `
+function overloadingThatCanReturnPromise(): Promise<number>;
+function overloadingThatCanReturnPromise(a: boolean): Promise<string>;
+async function overloadingThatCanReturnPromise(
+  a?: boolean,
+): Promise<number | string> {
+  return Promise.resolve(5);
+}
+      `,
+    },
+    {
+      code: `
+export function overloadingThatIncludeAny(): number;
+export function overloadingThatIncludeAny(a: boolean): any;
+export function overloadingThatIncludeAny(a?: boolean): any | number {
+  return Promise.resolve(5);
+}
```

Comment:
This and the test below aren't exactly valid TypeScript code (will also be flagged by `no-redundant-type-constituents`), but it did fail on my first pass at it (I've iterated the possible function signatures and their return types). Please let me know if I should drop these.